File: CoreDataKit/CoreDataKit.m
//
//  CoreDataKit.m
//  CoreDataKit
//
//  Created by Mathijs Kadijk on 15-05-14.
//  Copyright (c) 2014 Mathijs Kadijk. All rights reserved.
//

#import "CoreDataKit.h"
#import "CDKDebugger.h"
#import "NSPersistentStoreCoordinator+CoreDataKit.h"
#import "NSManagedObjectContext+CoreDataKit.h"

static NSString * const kCoreDataKitDefaultStoreName = @"CoreDataKit";

@interface CoreDataKit ()

@property (nonatomic, strong) NSPersistentStoreCoordinator *persistentStoreCoordinator;
@property (nonatomic, strong) NSManagedObjectContext *rootContext;
@property (nonatomic, strong) NSManagedObjectContext *mainThreadContext;

@end

@implementation CoreDataKit

+ (instancetype)sharedKit
{
    static CoreDataKit *sharedKit;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedKit = [[CoreDataKit alloc] init];
    });

    return sharedKit;
}

#pragma mark Setup methods

- (void)setupAutomigratingCoreDataStack
{
    [self setupCoreDataStack:kCoreDataKitDefaultStoreName
               automigrating:YES
          managedObjectModel:nil];
}

- (void)setupCoreDataStack:(NSString *)storeName
             automigrating:(BOOL)automigrating
        managedObjectModel:(NSManagedObjectModel *)managedObjectModelOrNil
{
    // Validate setup is only runned once
    NSAssert(nil == _rootContext, @"Root context is already available");
    NSAssert(nil == _mainThreadContext, @"Main thread context is already available");

    // Prep URL to location + get all models merged together
    NSURL *persistentStoreURL = [NSPersistentStoreCoordinator CDK_URLForStoreName:storeName];
    NSManagedObjectModel *mergedManagedObjectModel = (managedObjectModelOrNil) ?: [NSManagedObjectModel mergedModelFromBundles:nil];

    // Setup persistent store
    self.persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mergedManagedObjectModel];
    [self.persistentStoreCoordinator CDK_addSQLiteStoreWithURL:persistentStoreURL automigrating:automigrating];

    // Create the contexts
    self.rootContext = [NSManagedObjectContext CDK_contextWithPersistentStoreCoordinator:self.persistentStoreCoordinator];
    self.mainThreadContext = [self.rootContext CDK_childContextWithConcurrencyType:NSMainQueueConcurrencyType];
}

- (void)setupCoreDataStackInMemory
{
    [self setupCoreDataStackInMemoryWithManagedObjectModel:nil];
}

- (void)setupCoreDataStackInMemoryWithManagedObjectModel:(NSManagedObjectModel *)managedObjectModelOrNil
{
    // Validate setup is only runned once
    NSAssert(nil == _rootContext, @"Root context is already available");
    NSAssert(nil == _mainThreadContext, @"Main thread context is already available");

    // Get all models merged together
    NSManagedObjectModel *mergedManagedObjectModel = (managedObjectModelOrNil) ?: [NSManagedObjectModel mergedModelFromBundles:nil];
    
    // Setup persistent store
    NSError *error = nil;
    self.persistentStoreCoordinator = [[NSPersistentStoreCoordinator alloc] initWithManagedObjectModel:mergedManagedObjectModel];
    [self.persistentStoreCoordinator addPersistentStoreWithType:NSInMemoryStoreType configuration:nil URL:nil options:nil error:&error];

    CDKHandleError(error);

    // Create the contexts
    self.rootContext = [NSManagedObjectContext CDK_contextWithPersistentStoreCoordinator:self.persistentStoreCoordinator];
    self.mainThreadContext = [self.rootContext CDK_childContextWithConcurrencyType:NSMainQueueConcurrencyType];
}

#pragma mark Saving

+ (void)save:(CDKSaveBlock)saveBlock completion:(CDKCompletionBlock)completion
{
    [[self sharedKit] save:saveBlock completion:completion];
}

- (void)save:(CDKSaveBlock)saveBlock completion:(CDKCompletionBlock)completion
{
    NSManagedObjectContext *managedObjectContext = [self.rootContext CDK_childContextWithConcurrencyType:NSPrivateQueueConcurrencyType];
    [managedObjectContext performBlock:^{
        // Perform save block
        if (saveBlock) {
            saveBlock(managedObjectContext);
        }

        // Save the changes
        [managedObjectContext CDK_saveToPersistentStore:completion];
    }];
}

#pragma mark Accessors
- (NSPersistentStoreCoordinator *)persistentStoreCoordinator
{
    if (!_persistentStoreCoordinator) {
        CDKLogWarn(@"Persistent store coordinator on %@ is not set, did you forgot to call setup?", self);
    }

    return _persistentStoreCoordinator;
}

- (NSManagedObjectContext *)rootContext
{
    if (!_rootContext) {
        CDKLogWarn(@"Root context on %@ is not set, did you forgot to call setup?", self);
    }

    return _rootContext;
}

- (NSManagedObjectContext *)mainThreadContext
{
    if (!_mainThreadContext) {
        CDKLogWarn(@"Main thread context on %@ is not set, did you forgot to call setup?", self);
    }

    return _mainThreadContext;
}

@end

File: CoreDataKit/CoreDataKit.swift
//
//  CoreDataKit.swift
//  CoreDataKit
//
//  Created by Mathijs Kadijk on 23-06-14.
//  Copyright (c) 2014 Mathijs Kadijk. All rights reserved.
//

import CoreData

public enum LogLevel {
    case DEBUG
    case INFO
    case WARN
    case ERROR
}

public typealias Logger = (LogLevel, String) -> Void

/**
`CoreDataKit` helps with setup of the CoreData stack
*/
public class CoreDataKit : NSObject
{
    private struct Holder {
        static var sharedStack: CoreDataStack?
        static var sharedLogger: Logger = { _, message in println("[CoreDataKit] \(message)") }
    }

    /**
    Property to hold a shared instance of CoreDataKit, all the convenience class properties access and unwrap this shared instance. So make sure to set the shared instance before doing anything else.
    
    :discussion: This is the only property you have to set to setup CoreDataKit, changing the shared instace is not supported.
    */
    public class var sharedStack: CoreDataStack? {
        get {
            return Holder.sharedStack
        }

        set {
            Holder.sharedStack = newValue
        }
    }

    /**
    Shared logger used by CoreDataKit to log messages.
    
    :discussion: Default logger prints messages to console, but you can use this to use your own logger
    */
    public class var sharedLogger: Logger {
        get {
            return Holder.sharedLogger
        }

        set {
            Holder.sharedLogger = newValue
        }
    }

// MARK: Convenience properties

    /// Persistent store coordinator used as backing for the contexts of the shared stack
    public class var persistentStoreCoordinator: NSPersistentStoreCoordinator {
        return sharedStack!.persistentStoreCoordinator
    }

    /// Root context that is directly associated with the `persistentStoreCoordinator` and does it work on a background queue of the shared stack
    public class var rootContext: NSManagedObjectContext {
        return sharedStack!.rootContext
    }

    /// Context with concurrency type `NSMainQueueConcurrencyType` for use on the main thread of the shared stack
    public class var mainThreadContext: NSManagedObjectContext {
        return sharedStack!.mainThreadContext
    }

    /**
    Creates a child context with the root context of the shared stack as parent and performs the given block on the created context.

    :param: block       Block that performs the changes on the given context that should be saved
    :param: completion  Completion block to run after changes are saved

    :see: NSManagedObjectContext.performBlock()
    */
    public class func performBlockOnBackgroundContext(block: PerformBlock, completionHandler: PerformBlockCompletionHandler? = nil) {
        sharedStack!.performBlockOnBackgroundContext(block, completionHandler: completionHandler)
    }
}
